crates/trie/trie/src/proof_v2/mod.rs

```diff
@@ -570,27 +646,454 @@ where
+        level = "trace",
+        skip(self, value_encoder, targets, hashed_cursor_current),
+    )]
+    fn calculate_key_range(
+        &mut self,
+        value_encoder: &VE,
+        targets: &mut TargetsIter<impl Iterator<Item = Nibbles>>,
+        hashed_cursor_current: &mut Option<(Nibbles, VE::DeferredEncoder)>,
+        lower_bound: Nibbles,
+        upper_bound: Option<Nibbles>,
+    ) -> Result<(), StateProofError> {
+        // A helper closure for mapping entries returned from the `hashed_cursor`, converting the
+        // key to Nibbles and immediately creating the DeferredValueEncoder so that encoding of the
+        // leaf value can begin ASAP.
+        let map_hashed_cursor_entry = |(key_b256, val): (B256, _)| {
+            debug_assert_eq!(key_b256.len(), 32);
+            // SAFETY: key is a B256 and so is exactly 32-bytes.
+            let key = unsafe { Nibbles::unpack_unchecked(key_b256.as_slice()) };
+            let val = value_encoder.deferred_encoder(key_b256, val);
+            (key, val)
+        };
+
+        // If the cursor hasn't been used, or the last iterated key is prior to this range's
+        // key range, then seek forward to at least the first key.
+        if hashed_cursor_current.as_ref().is_none_or(|(key, _)| key < &lower_bound) {
+            let lower_key = B256::right_padding_from(&lower_bound.pack());
+            *hashed_cursor_current =
+                self.hashed_cursor.seek(lower_key)?.map(map_hashed_cursor_entry);
+        }
+
+        // Loop over all keys in the range, calling `push_leaf` on each.
+        while let Some((key, _)) = hashed_cursor_current.as_ref() &&
+            upper_bound.is_none_or(|upper_bound| key < &upper_bound)
+        {
+            let (key, val) =
+                core::mem::take(hashed_cursor_current).expect("while-let checks for Some");
+            self.push_leaf(targets, key, val)?;
+            *hashed_cursor_current = self.hashed_cursor.next()?.map(map_hashed_cursor_entry);
+        }
+
+        Ok(())
+    }
+
+    /// Constructs and returns a new [`ProofTrieBranch`] based on an existing [`BranchNodeCompact`].
+    #[inline]
+    const fn new_from_cached_branch(
+        cached_branch: &BranchNodeCompact,
+        ext_len: u8,
+    ) -> ProofTrieBranch {
+        ProofTrieBranch {
+            ext_len,
+            state_mask: TrieMask::new(0),
+            masks: TrieMasks {
+                tree_mask: Some(cached_branch.tree_mask),
+                hash_mask: Some(cached_branch.hash_mask),
+            },
+        }
+    }
+
+    /// Pushes a new branch onto the `branch_stack` which is based on a cached branch obtained via
+    /// the trie cursor.
+    ///
+    /// If there is already a child at the top branch of `branch_stack` occupying this new branch's
+    /// nibble then that child will have its short-key split with another new branch, and this
+    /// cached branch will be a child of that splitting branch.
+    fn push_cached_branch(
```

Comment:
This method is the sibling of `push_leaf`. Where `push_leaf` was used to handle all changes required to push a single leaf onto the `child_stack`, this method handles all changes required to push a single cached branch onto the `branch_stack`.